src/conn/conn_compact.c

```diff
@@ -0,0 +1,253 @@
+        if (ret == EBUSY || ret == WT_NOTFOUND || ret == ETIMEDOUT)
+            ret = 0;
+        WT_ERR(ret);
+    }
+
+    WT_STAT_CONN_SET(session, session_background_compact_running, 0);
+
+    WT_ERR(__wt_metadata_cursor_close(session));
+    __wt_free(session, config);
+    __wt_free(session, conn->background_compact.config);
+
+    if (0) {
+err:
+        WT_IGNORE_RET(__wt_panic(session, ret, "compact server error"));
+    }
+    return (WT_THREAD_RET_VALUE);
+}
+
+/*
+ * __wt_compact_server_create --
+ *     Start the compact thread.
+ */
+int
+__wt_compact_server_create(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    uint32_t session_flags;
+
+    conn = S2C(session);
+
+    /* Set first, the thread might run before we finish up. */
+    FLD_SET(conn->server_flags, WT_CONN_SERVER_COMPACT);
+
+    /*
+     * Compaction does enough I/O it may be called upon to perform slow operations for the block
+     * manager.
+     */
+    session_flags = WT_SESSION_CAN_WAIT;
+    WT_RET(__wt_open_internal_session(
+      conn, "compact-server", true, session_flags, 0, &conn->background_compact.session));
+    session = conn->background_compact.session;
+
+    WT_RET(__wt_cond_alloc(session, "compact server", &conn->background_compact.cond));
+
+    WT_RET(__wt_thread_create(session, &conn->background_compact.tid, __compact_server, session));
+    conn->background_compact.tid_set = true;
+
+    return (0);
+}
+
+/*
+ * __wt_compact_server_destroy --
+ *     Destroy the background compaction server thread.
+ */
+int
+__wt_compact_server_destroy(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+
+    conn = S2C(session);
+
+    FLD_CLR(conn->server_flags, WT_CONN_SERVER_COMPACT);
+    if (conn->background_compact.tid_set) {
+        conn->background_compact.running = false;
+        __wt_cond_signal(session, conn->background_compact.cond);
+        WT_TRET(__wt_thread_join(session, &conn->background_compact.tid));
+        conn->background_compact.tid_set = false;
+    }
+    __wt_cond_destroy(session, &conn->background_compact.cond);
+
+    /* Close the server thread's session. */
+    if (conn->background_compact.session != NULL) {
+        WT_TRET(__wt_session_close_internal(conn->background_compact.session));
+        conn->background_compact.session = NULL;
+    }
+
+    return (ret);
+}
+
+/*
+ * __wt_compact_signal --
+ *     Signal the compact thread. Return EBUSY if the background compaction server has not processed
+ *     a previous signal yet.
+ */
+int
+__wt_compact_signal(WT_SESSION_IMPL *session, const char *config)
+{
+    WT_CONNECTION_IMPL *conn;
+    bool running;
+
+    conn = S2C(session);
+
+    /* Wait for any previous signal to be processed first. */
+    if (conn->background_compact.signalled)
+        return (EBUSY);
+
+    __wt_spin_lock(session, &conn->background_compact.lock);
+    running = conn->background_compact.running;
+    conn->background_compact.running = !running;
```

Comment:
The main reason behind it is to stop the background compact and then change the config. But this doesn't turn it to running again if the background compact is running already.

Reply:
If the application executes:

```
session->compact("background=true")
session->compact("background=true")
```

It will lead to the background server stopping which is wrong. Is that what you mean? I am not sure where it is best to make the second call return an error. Here or in the `session::compact` function?

I guess here would be the best?

Reply:
I have added a FIXME comment as WT-11334 should take care of the configuration parsing.
cc @Sean04

Reply:
Yes, that is a scenario that I have in mind. Do we have any way to identify whether the background compact is running/paused? The error message to the user must be clear that the background compact thread is already running.

Reply:
Yes, the value of `conn->background_compact.running` should indicate the state of the background compaction server. The function

Reply:
The parsing was not working correctly after further testing, more changes in ddc9e914caa4239ceca35ab7b5021da742be7205.